# RARE
A framework for random region generation  

National Center for Tumor Diseases (NCT) Heidelberg  
Section Molecular and Gene Therapy  
Author: Elvira Mingazova  
Day-to-day supervisor: Dr. Raffaele Fronza  
Start 06.03.17  

This GitHub Repository contains all the files on the Bachelor project developed by me during my internship at NCT Heidelberg. The task was to develop three models for random region generation:
1. Random Model Zero (RM0): unconditional random IS generation
2. Random Model One (RM1): mappability
3. Random Model Two (RM2): enzyme bias  

The datasets generated by the listed models were used by my supervisor in the context of a bigger integration site (IS) analysis on the mutational toxicity of viral vector based gene therapy. *In silico* generated IS were compared with the experimental IS in order to see whether the experimental integration landscape matches the random one.  

The code was written and tested mainly in the Jupyter Notebook and contains parts in Python (version 2.7.13) and Bash. The notebook files are located in the folder *Notebooks*. If you will go through them in the chronological order, you will be able to track how exactly each of the problems of creating the three models of RARE was solved by me. The chronological order is the following:  
- **getRegionW1.ipynb:** RM0 the simplest version
- **getRegionW2.ipynb:** RM0 with the testing of the input file features, explanation of the differences between two major human genome annotations (hg38 and hg19) and IS distribution plots
- **Kolmogorov-Smirnov-Test.ipynb:** theory and visualization of the Kolmogorov-Smirnov statistical two-sample test for the later use in the comparison of the datasets generated by three different random models.
- **genComparison.ipynb:** some thoughts on the mappability expectations for three different DNA sequences: E.Coli whole genome, Yeast whole genome and human chromosome Y.
- **mappabilityTrack.ipynb:** the concept of mappability, explanation of how to use GEM-tools in order to compute the mappability of the genome of choice and visualize the mappability track using IGV and prove of the correctness of the results
- **ModelOne_Mappability.ipynb:** explaination of how to arrive from the precomputed mappability file to the generation of IS on the uniquely mappable regions of the respective genome, distribution of the IS generated by RM1 on chr1 as an example
- **ModelTwo_EnzymeBias.ipynb:** generation of the restriction enzyme sites matched IS on the uniquely mappable regions on the genome of choice, distribution of the IS generated with RM2 on chr1, Kolmogorov-Smirnov tests to compare three random models.

The next step would be to create a pipeline and automatize the process of random regions generation.

*Please note that this project was developed by a bachelor biochemistry student curious to try herself in bioinformatics, so the way the problems are solved may be to a certain extend far from the optimal*
